### PR TITLE
Add missing margin to SnackBarAction

### LINKED ISSUE #48042: SnackBarAction doesn't have right margin
image

As you can see above, the `SnackBarAction` does not have any space between right button border and right snackbar border which is not a correct appearance according to Material Design guidelines: https://material.io/components/snackbars/#specs

## Steps to Reproduce

```dart
Scaffold.of(context).showSnackBar(
  SnackBar(
    behavior: SnackBarBehavior.floating,
    duration: Duration(days: 1),
    content: Text('Usunięto element z listy.'),
    action: SnackBarAction(
      textColor: Color(0xFFFFF0C3),
      label: 'COFNIJ',
      onPressed: () {},
    ),
  ),
);
```

```
[√] Flutter (Channel dev, v1.13.5, on Microsoft Windows [Version
    10.0.18362.535], locale pl-PL)
    • Flutter version 1.13.5 at D:\flutter
    • Framework revision 41a911099b (13 days ago), 2019-12-19 13:48:02     
      -0800
    • Engine revision 0f90e6546b
    • Dart version 2.8.0 (build 2.8.0-dev.0.0 aa6709974d)


[√] Android toolchain - develop for Android devices (Android SDK version   
    28.0.3)
    • Android SDK at C:\Users\walbe\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native     
      profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java 
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 40.0.2
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b03)

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 9  
      (API 28) (emulator)
```

### PR DESCRIPTION
**Pull request by Albert221**
## Description

Fixes #48042. Adds a horizontal margin to the SnackBar action, according to [Material Design guidelines](https://material.io/components/snackbars#specs).

As you can see on the first screenshot below (**before**), the SnackBarAction does not have any space between the right button border and the right snackbar border which is not a correct appearance according to Material Design guidelines

### Before

image

### After

image
image

## Related Issues

#48042

## Tests

I added the following tests:

- SnackBar has correct end padding when it contains an action with fixed behavior
- SnackBar has correct end padding when it contains an action with floating behavior

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change.
   - [ ] I wrote a design doc: https://flutter.dev/go/template 
   - [ ] I got input from the developer relations team, specifically from: 
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template

Yes, I changed some of the existing tests, but they weren't following the spec in Material Design guidelines too. That change is really cosmetic, I don't think there is a need for a migration guide or a design doc.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/snack_bar.dart [hunk 1]
```diff
@@ -428,6 +428,8 @@ class _SnackBarState extends State<SnackBar> {
     final EdgeInsetsGeometry padding = widget.padding
       ?? EdgeInsetsDirectional.only(start: horizontalPadding, end: widget.action != null ? 0 : horizontalPadding);
 
+    final double actionHorizontalMargin = (widget.padding?.resolve(TextDirection.ltr).right ?? horizontalPadding) / 2;
```
Comment (Piinks) [inline packages/flutter/lib/src/material/snack_bar.dart:431]: How did you calculate this? It looks like the spec shows a standard 8 pixels as the margin: https://material.io/components/snackbars#specs

**Review by Piinks — COMMENTED** [review]
Hi @Albert221 welcome! And thank you for the contribution. It looks like there is a merge conflict, can you update your change to resolve it?

File: packages/flutter/lib/src/material/snack_bar.dart (hunk 1 above)
Comment (Albert221) [reply packages/flutter/lib/src/material/snack_bar.dart:431]: This was a little bit tricky. Spec indeed shows 8 pixels of margin here, but spec is only for a floating SnackBar. In the `SnackBar` we have a `padding` parameter which allows the developer to customize it. I figured out that in the spec, left padding of the whole snackbar is 16 px and right padding is 8px. But there is also a left padding for the button itself that's the same as this right padding - 8px. 16px on the left equals to 8px + 8px on the right. The screenshot below describes my idea behind that.

image

Thanks to that even after providing a custom `padding` by the developer, our spacing will still be proportional to what the spec shows! :)

Comment (Piinks) [reply packages/flutter/lib/src/material/snack_bar.dart:430]: Awesome! Thank you for the explanation. :)

**Comment by Piinks** [comment]
It looks like the latest update included the NNBD migration of the test files you are working in, so your tests will need to be null safe. That should resolve the test failures here.

**Comment by Albert221** [comment]
Of course, here's the fixed version!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

**Pull request merged**